Added option to not parse data

gabrielecoppi opened the pull request:
# pyubx2 Pull Request Template

## Description

Added the option in `UBXReader.read()` to not parse the payload. This may be useful when continuously polling data with a low power computer and a lot of messages requested.

Fixes # (issue)

## Testing

Added two tests with the flag `parsing=False` to verify the behavior and they work (NAVHPPOS and HNRLOG) in `test_stream.py`

## Checklist:

- [ x] I agree to abide by the code of conduct (see [CODE_OF_CONDUCT.md](https://github.com/semuconsulting/pyubx2/blob/master/CODE_OF_CONDUCT.md)).
- [ x] My code follows the style guidelines of this project (see [CONTRIBUTING.MD](https://github.com/semuconsulting/pyubx2/blob/master/CONTRIBUTING.md)).
- [ x] I have performed a self-review of my own code.
- [ x] (*if appropriate*) I have cited my u-blox documentation source(s).
- [ x] I have commented my code, particularly in hard-to-understand areas.
- [ x] I have made corresponding changes to the documentation.
- [ x] (*if appropriate*) I have added test cases to the `tests/test_*.py` unittest suite to maintain >= 99% code coverage.
- [ x] I have tested my code against the full `tests/test_*.py` unittest suite.
- [ x] My changes generate no new warnings.
- [ x] Any dependent changes have been merged and published in downstream modules.
- [ x] I have [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) my commits.
- [ x] I understand and acknowledge that the code will be published under a BSD 3-Clause license.

semuadmin commented:
Hi @gabrielecoppi 

Thanks for this, but I'm a little unclear on the practical utility? pyubx2 is a protocol parser - I'm happy to be convfinced otherwise, but removing the parsing option feels a bit like buying a car and taking the wheels off. WIthout the parsed data, you're simply reading the raw binary datastream, and you can do that with any serial reader utility - you wouldn't need pyubx2 at all.

gabrielecoppi commented:
Hi @semuadmin, I give you my specific application (for which I have modified `pyubx2`). The package is still parsing by default but now I have added the option to not do it. I am currently using a RPi that is powered by a battery and it monitors multiple devices including a ublox. I found that I save some CPU without parsing data (and so I get the battery to last a bit longer). I also found out that with this structure I can save the message with the correct length, instead of random chunks of bytes. This ensures that all the messages saved are correct and significantly simplify the analysis whit the data. Finally, in my specific application saving binary format helps in saving space on the SD of the RPi.

semuadmin commented:
> Hi @semuadmin, I give you my specific application (for which I have modified `pyubx2`). The package is still parsing by default but now I have added the option to not do it. I am currently using a RPi that is powered by a battery and it monitors multiple devices including a ublox. I found that I save some CPU without parsing data (and so I get the battery to last a bit longer). I also found out that with this structure I can save the message with the correct length, instead of random chunks of bytes. This ensures that all the messages saved are correct and significantly simplify the analysis whit the data. Finally, in my specific application saving binary format helps in saving space on the SD of the RPi.

OK no issue with the change in principle but, as I say, if you simply want to save the binary UBX data without the CPU overhead of parsing, why not just pipe the serial stream directly to a file e.g. something like:

```
stdbuf -o0 cat /dev/ttyACM0 > ubx.bin
```


If you need to delimit individual UBX messages, you could grep the UBX delimiter b'\xb5\x62'. Just a thought...

gabrielecoppi commented:
I agree. This would be easier, but in my mind I am building this system in a future proof way, such that if I change the RPi to a different board, if I move to more fixed system with wall power, I just need to change a single flag and the code still work with the plus that now I can have the data parsed. 
I also forgot to mention that at the moment based on the measurement that I want to perform I need to configure the ublox with different output and that is done at the startup of the RPi, so I need to use a VALSET command (if I want or not the RTK data for example).
As I mentioned is a specific case of mine, so I do not want to change the paradigm of the code that you wrote (thanks a lot by the way), I just thought that since I have done this modification and this is not affecting anything may be useful for others too

semuadmin commented on .vscode/settings.json:
```diff
@@ -1,11 +1,15 @@
     "moduleversion": "1.2.25",
-    "sarif-viewer.connectToGithubCodeScanning": "off"
+    "sarif-viewer.connectToGithubCodeScanning": "off",
+    "[python]": {
```
You are of course at liberty to use your own VSCode workflow, but I'd prefer to avoid making changes to the one provided, and it's not integral to the PR. Can we remove these changes please?

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -109,23 +109,22 @@ def __next__(self) -> tuple:
         return (raw_data, parsed_data)
 
-    def read(self) -> tuple:
+    def read(self, parsing=True) -> tuple:
```
This works (more or less) if you're invoking the read() method directly, but it doesn't work if you're using the iterator. It might be better to make this a UBXReader constructor keyword argument and carry it through from there.

semuadmin replied:
This causes the program to hang if you're reading from a dummy socket input (via pyubx2's interal SocketStream wrapper) - try running /tests/test_socket.py.

semuadmin commented on tests/test_stream.py:
```diff
@@ -173,6 +173,18 @@ def testHNRLOG(
                 self.assertEqual(str(parsed), EXPECTED_RESULTS[i])
```
You've ticked the checkbox in the PR template, but did you actually run your changes through the entire unittest suite? I'm seeing quite a few failures in /tests/test_stream.py, and /tests/test_socket.py hangs completely.

<img width="358" alt="Screenshot 2023-05-10 at 19 29 03" src="https://github.com/semuconsulting/pyubx2/assets/28569967/b34ba1fe-ed08-4fcf-8c50-fb622d09270d">

semuadmin reviewed (changes requested):
Happy in principle but implementation needs more work - see comments

gabrielecoppi commented:
Ok I will reply here. I apologize for messing up the VScode settings. 
For the other comments as suggested I set a `self._parsing` constructor in `UBXReader` to not break the iterator. The issues with the tests was that I messed up with the bitwise operator `&` when checking the protocol filter. Now, it should be solved. I created a separate testing file for a no parsing case. Everything seems to work as can be seen
![image](https://github.com/semuconsulting/pyubx2/assets/29742144/35dd4a2a-8d37-46ff-a0bc-3edc82e42d3b)

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -84,6 +84,8 @@ def __init__(self, datastream, **kwargs):
         self._labelmsm = int(kwargs.get("labelmsm", True))
         self._msgmode = int(kwargs.get("msgmode", GET))
 
+        self._parsing = kwargs.get("parsing", True)
```
this is fine, but need to update docstring at the top of the module:

:param bool parsing: a boolean flag to determine if the stream needs to be parsed

(you can remove this docstring entries from the parse_* methods as it no longer applies)

gabrielecoppi replied:
Added

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -117,15 +119,16 @@ def read(self) -> tuple:
         'protfilter' determines which protocols are parsed.
         'quitonerror' determines whether to raise, log or ignore parsing errors.
 
+        :param bool parsing: a boolean flag to determine if the stream needs to be parsed
```
can remove this docstring entry as it no longer applies

gabrielecoppi replied:
Removed

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -197,6 +200,7 @@ def _parse_ubx(self, hdr: bytes) -> tuple:
         Parse remainder of UBX message.
 
         :param bytes hdr: UBX header (b'\xb5\x62')
+        :param bool parsing: a boolean flag to determine if the stream needs to be parsed
```
can remove this docstring entry as it no longer applies

gabrielecoppi replied:
Removed

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -230,6 +234,7 @@ def _parse_nmea(self, hdr: bytes) -> tuple:
         Parse remainder of NMEA message (using pynmeagps library).
 
         :param bytes hdr: NMEA header ($G or $P)
+        :param bool parsing: a boolean flag to determine if the stream needs to be parsed
```
can remove this docstring entry as it no longer applies

gabrielecoppi replied:
Removed

semuadmin commented on src/pyubx2/ubxreader.py:
```diff
@@ -257,6 +262,7 @@ def _parse_rtcm3(self, hdr: bytes) -> tuple:
         Parse any RTCM3 data in the stream (using pyrtcm library).
 
         :param bytes hdr: first 2 bytes of RTCM3 header
+        :param bool parsing: a boolean flag to determine if the stream needs to be parsed
```
can remove this docstring entry as it no longer applies

gabrielecoppi replied:
Removed

semuadmin commented on tests/test_stream.py:
```diff
@@ -540,7 +540,6 @@ def testUBXITERATE_ERR2(
         i = 0
         for raw, parsed in ubr:
             if not isinstance(parsed, str):
```
good spot, but I've actually been a bit lazy with this particular test case and should have validated the stdout output rather than just count the lines. I've since updated this in the master branch.

semuadmin commented on tests/test_stream_no_parsing.py:
```diff
@@ -0,0 +1,514 @@
+"""
```
these test cases need revising and simplifying. You need to check the actual output, not just validate that the number of lines is correct (_I appreciate you probably picked this up from one of my own lazier test cases! - see comment above_). You'll need to update the EXPECTED_RESULTS entries to reflect the raw binary output, but you don't need to go overboard - just one test case per protocol (NMEA, UBX. RTCM), or if you prefer a single test case combining all 3 protocols. That will certainly make the DIFF a lot smaller!

gabrielecoppi replied:
I did update the no parsing test for both the iterator and the read method for a mixed messages (RTCM, UBX and NMEA)

gabrielecoppi commented:
Ok I did not have noticed that you updated the `test_stream.py`. So I have updated my branch too, to reflect this
